File: libraries/ary/ary1_id2ac.f
      SUBROUTINE ARY1_ID2AC( IARY, IACB )
*+
*  Name:
*     ARY1_ID2AC

*  Purpose:
*     Convert an array identifier into the associated ACB index.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL ARY1_ID2AC( IARY, IACB )

*  Description:
*     The routine converts an array identifier, previously issued by
*     ARY1_EXPID, into an index to the appropriate entry in the ACB.
*     The identifier supplied is fully checked and a value of zero is
*     returned if it is not valid.

*  Arguments:
*     IARY = INTEGER (Given)
*        Array identifier.
*     IACB = INTEGER (Returned)
*        Index to an entry in the ACB, or zero of the identifier
*        supplied was not valid.

*  Notes:
*     -  This routine does not perform error checking or reporting.

*  Algorithm:
*     -  Check that the IARY value supplied is positive and return a
*     value of zero if it is not.
*     -  Decode the IARY value into an index for the ACB.
*     -  Check that the identifier value matches the value originally
*     issued for the ACB slot and that the slot is still in use.
*     -  If everything is OK, then return the ACB index. Otherwise,
*     return zero.

*  Copyright:
*     Copyright (C) 1989 Science & Engineering Research Council.
*     All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 59 Temple Place,Suite 330, Boston, MA
*     02111-1307, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     {enter_new_authors_here}

*  History:
*     31-JUL-1989 (RFWS):
*        Original version.
*     22-AUG-1989 (RFWS):
*        Added EXTERNAL statement to ensure ARY1_INIT is linked.
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-

*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'ARY_PAR'          ! ARY_ public constants
      INCLUDE 'ARY_CONST'        ! ARY_ private constants

*  Global Variables:
      INCLUDE 'ARY_ACB'          ! ARY_ Access Control Block
*        ACB_CHK( ARY__MXACB ) = INTEGER (Read)
*           Identifier value originally issued for each ACB slot.
*        ACB_USED( ARY__MXACB ) = LOGICAL (Read)
*           Whether the ACB slot is in use.

*  Arguments Given:
      INTEGER IARY

*  Arguments Returned:
      INTEGER IACB

*  External References:
      EXTERNAL ARY1_INIT         ! Initialise common blocks

*  Local Variables:
      INTEGER I                  ! Temporary variable for ACB index

*.

*  Check that the identifier value is positive, and return a value of
*  zero if it is not.
      IF ( IARY .LE. 0 ) THEN
         IACB = 0

*  Decode the IARY value into an index for the ACB (the reverse of the
*  process used by ARY1_EXPID to encode it).
      ELSE
         I = MOD( IARY, ARY__MXACB )
         IF ( I .EQ. 0 ) I = ARY__MXACB

*  Check that the identifier matches the value originally issued for
*  this ACB slot and that the slot is still in use. If OK, then return
*  the ACB index. Otherwise, return zero.
         IF ( ( ACB_CHK( I ) .EQ. IARY ) .AND. ACB_USED( I ) ) THEN
            IACB = I
         ELSE
            IACB = 0
         END IF
      END IF

      END

File: libraries/cary/ary1Id2ac.c
#include <pthread.h>
#include "sae_par.h"
#include "ary1.h"

/* Define a union which allows a bit pattern to be accessed as a
   signed or unsigned int, or as a pointer. */
typedef union IdUnion {
   int i;
   unsigned u;
   const void *pointer;
} IdUnion;

/* These globals are declared in ary1Ffs.c */
extern int Ary_NACB;
extern int Ary_NPCB;
extern AryACB **Ary_ACB;
extern AryPCB **Ary_PCB;

AryObject *ary1Id2ac( const void *id_ptr, int isacb ) {
/*
*+
*  Name:
*     ary1Id2Ac

*  Purpose:
*     Convert an identifier into the associated ACB or PCB pointer.

*  Synopsis:
*     AryObject *ary1Id2ac( const void *id_ptr, int isacb )

*  Description:
*     The routine converts an identifier, previously issued by
*     ary1Expid, into a pointer to the appropriate ACB pr PCB structure.
*     The identifier supplied is fully checked and a NULL pointer is
*     returned if it is not valid.

*  Parameters:
*     id_ptr
*        The identifier, in the form of an Ary or AryPlace pointer.
*     isacb
*        Indicates the type of object expected by the calling function.
*        Non-zero indicates an ACB is expected. Zero indicates a PCB is
*        expected.

*  Returned function value:
*     Pointer to the ACB or PCB structure, or NULL of the identifier supplied
*     was not valid.

*  Notes:
*     -  This routine does not perform error checking or reporting.

*  Copyright:
*      Copyright (C) 2017 East Asian Observatory
*      All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     03-JUL-2017 (DSB):
*        Original version, based on AST identifier system by RFWS.
*     4-SEP-2017 (DSB):
*        Modified to handle either ACB or PCB identifiers.
*-
*/

/* Local variables: */
   AryObject *result = NULL;
   IdUnion work;
   int id;
   int slot;
   int nel;

/* Check an identifier was supplied. */
   if( !id_ptr ) return result;

/* Wait for exclusive access to the ACB or PCB related global variables */
   if( isacb ) {
      ARY__ACB_LOCK_MUTEX;
   } else {
      ARY__PCB_LOCK_MUTEX;
   }

/* If OK, reverse the encoding process performed by ary1Expid to
   retrieve the slot index for the ACB. */
   work.pointer = id_ptr;
   id = work.i;
   work.u = ( work.u ^ ( ( (unsigned) ARY__FACNO ) << 8U ) ) >> 8U;

/* Check that the offset obtained doesn't extend beyond the limits of
   the array of ACB pointers. */
   nel = isacb ? Ary_NACB : Ary_NPCB;
   if ( ( work.i >= 0 ) && ( work.i < nel ) ) {

/* Get a pointer to the ACB structure. Remember that ary1Expid converts
   the slot number from zero-base to one-base, so we need to convert it
   back to zero-base before using it. */
      slot = work.i - 1;
      if( isacb ) {
         result = (AryObject *) Ary_ACB[ slot ];
      } else {
         result = (AryObject *) Ary_PCB[ slot ];
      }

/* See if the "check" field matches the ID value. */
      if( result->check != id ) {
         result = NULL;

/* Also check that the slot number stored in the object is the expected
   value. */
      } else if( result->slot != slot ) {
         result = NULL;
      }
   }

/* Allow other threads to access the ACB or PCB related global variables */
   if( isacb ) {
      ARY__ACB_UNLOCK_MUTEX;
   } else {
      ARY__PCB_UNLOCK_MUTEX;
   }

/* Return the result. */
   return result;
}


